### ISSUE TITLE
Feature request: Configuration option to disable generation of `[k: string]: unknown`

### ISSUE THREAD

**Issue by binaryunary** [bcherny/json-schema-to-typescript#335]
Hi,

Would you consider adding a config option to disable generation of `[k: string]: unknown;` if `additionalProperties` is not explicitly set to false?

I understand that technically the current behavior is correct as it follows the spec but `[k: string]: unknown;` allows referencing any arbitrary properties and makes it difficult to catch errors at compile time.

**Comment by tadhgmister** [comment]
you definitely want to complain to who ever owns the schema telling them to add `additionalProperties: false` since users of the schema that don't use typescript will face the same issue.

**Comment by bcherny** [comment]
Agreed with @tadhgmister -- while this may seem appealing, it's an easy way to introduce subtle bugs where codegenned typings behave differently for different languages. We want to stick as closely as possible to the JSON-Schema spec to make the compiler predictable and reliable.

Leaving this open for now to collect more feedback.

**Comment by johnbillion** [comment]
Agreed that this is a problem that should be fixed in the JSON schema, not at the point where its transformed into TS interfaces. If the JSON schema you're referencing is used for validation then the "any arbitrary properties" you refer to are perfectly valid.

**Comment by Karel-Kroeze** [comment]
while I agree that forcing `additionalProperties` to be false should never be the default, having the option to do so through an option for overly permissive schemas would be extremely handy.

**Comment by dhoepelman** [comment]
I would like to vote for this as a config option.
I'd even argue that it should be the default as typescript programs can always have excess properties at runtime¹.

`additionalProperties: false` has some caveats for use - adding a field in a newer schema becomes a breaking change - and you don't always control the schema.

You should still exclusively use known properties in your programs. `[k: string]: unknown` makes that more error-prone.

¹ Typescript uses duck typing. [It can prevent additional properties in special cases](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks).

Example ([Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAbgQwDYFcCmMC8MDeAoGQmBALhgHIFyAaAogIzPPvLwF888AzFMYKAJbgYXABQCwABxRQyOYmWgAnCQHMYbAJS46hJWigolYEcghoA3O05jEqNNoD0jmGBAw0SpSCVA))
```typescript
const value = {
    a: 'a',
    b: 'b'
}

function f(input: { a: string }) {
    return false;
}

f(value) // no error
```

**Comment by mgabeler-lee-6rs** [comment]
A reason to make this an option: TypeScript _already_ treats all types as `additionalProperties: true` in _many_ contexts, and you cannot suppress this AFAICT.

Fundamentally, TypeScript types and interfaces specify what a value must _have_, not what it must _not_ have. Thus this is perfectly fine:
```typescript
export interface Foo {
    bar: number;
    baz: string;
}
const x = {
    bar: 12,
    baz: "13",
    quux: [14],
};
const y: Foo = x;
const yy: Foo = { ...x };

const z = {
    bar: 12,
    baz: "13",
    quux: [14],
} as Foo;
const zz = { ...x } as Foo;
```
[Playground Link](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgMQgjgG8AoOSuAI0ygC44kBXAW2rQG4KraAvRgGcYUZAHNuAXzLYISYXBBwAvKR6VaDOAEYATABp1NTALgAibQGYzhqnACOzZiEYBtbQBYAuocndZ8vAAnoyExKog-nIKQSEERCqkcAB0qUp+ZDLR8HyJ5HaajHq2vCaMFtYllI7Obp4+ZJJwmILxEFGBcHy5qiQpaXBNLW3cQA)

The only thing in this area where TypeScript would complain is this pattern:
```typescript
const a: Foo = {
    bar: 12,
    baz: "13",
    quux: [14],
}
// Type '{ bar: number; baz: string; quux: number[]; }' is not assignable to type 'Foo'.
//   Object literal may only specify known properties, and 'quux' does not exist in type 'Foo'.
```

----

Second rationale: If you use `allOf` to combine schemas, you _cannot_ set `additionalProperties: false`. If you do that, you generally end up with schemas which exclude all possible documents, because you e.g. have "schema `X` says must have props `a` and `b` and nothing else, schema `Y` says `c` and `d` and nothing else, so `allOf[X,Y]` must have `a`, `b`, `c`, and `d`, while also simultaneously _**having no properties at all**_". The response from the JSON schema folks to this problem seems to be, roughly, "we don't support inheritance".

This issue results in it being impossible for my team to have `additionalProperties: false` in almost any of our schemas.

**Comment by nbrahms** [comment]
👋 Hi folks,

I recently started testing out this tool to generate typescript types from a json-schema dependency in a project I'm working on.

I want to +1 this request and explain _why_ we need it:

It's worth asking what the purpose of a `jsonschema` is. It can be used for multiple needs. In particular, it's most often used for _data validation_. In this case, `additionalProperties: false` tells the data validator that any unknown object property should result in a validation error.

In the case of `json-schema-to-typescript`, the `jsonschema` is more likely (and is, in my specific use case) used for _code generation_; viz., we want to make sure we don't write TypeScript code that accesses a missing field. Using `json-schema-to-typescript`'s current behavior, we would have to set `additionalProperties: false` to prevent unbounded field access to objects.

Now let's consider creating a forward-compatible TypeScript consumer of an API, where the API is described by a `jsonschema`. In order to allow extension of the API in a forward-compatible manner, we should leave `additionalProperties` unset, to prevent schema validation errors. But in order to catch backward-incompatible changes in the schema in our TypeScript code, we need the generated type definitions to not include unbounded field accessors (i.e., `[k: string]: unknown`).

All the above is given one even has control over the jsonschema in question.

So TL;DR: I'd use `json-schema-to-typescript` if it has this requested configuration option; without it, it's a tough sell to my team to adopt.

**Comment by bcherny** [comment]
So while I'm not super enthused about this feature, it seems like a good number of others are. The next JSTT release will include a configurable `additionalProperties` option, which you can default to `true` or `false` (`true` by default, per the spec).

**Comment by nbrahms** [comment]
Thanks @bcherny! Appreciate the update.

**Comment by bcherny** [comment]
Published as part of v11.0.0.

**Comment by john30** [comment]
thanks for adding this feature @bcherny 
just tried it and it seems that the call to minimist misses declaration of --additionalProperties being boolean as e.g. using "--additionalProperties false" yields in additionalProperties='false' being used as default instead of additionalProperties=false

**Comment by bcherny** [comment]
@john30 I think you're missing the `=`: `--additionalProperties=false`. https://github.com/bcherny/json-schema-to-typescript/commit/413517d6498428f997cf8de29d937cf895c1fed0

**Comment by john30** [comment]
no, the type is just kept as string without my proposed patch and as such 'false' evaluates trueish

**Comment by bcherny** [comment]
@john30 I don't understand what you mean. See the commit I linked, showing off the behavior. Does the output on your machine differ from the output in the commit? Or are you using a different command?

**Comment by john30** [comment]
giving your linked example a try with "--additionalProperties=false" (without my patch) yields in the following result:
```typescript
export interface ExampleSchema {
  firstName: string;
  lastName: string;
  /**
   * Age in years
   */
  age?: number;
  height?: number;
  favoriteFoods?: unknown[];
  likesDogs?: boolean;
  [k: string]: "false";
}
```
where it actually should be this (with my patch applied):
```typescript
export interface ExampleSchema {
  firstName: string;
  lastName: string;
  /**
   * Age in years
   */
  age?: number;
  height?: number;
  favoriteFoods?: unknown[];
  likesDogs?: boolean;
}
```

this is caused by minimist not knowing that the argument "additionalProperties" is of type boolean and as such it interprets it as string giving the extra (unexpected) declaration of
`[k: string]: "false";`

**Comment by john30** [comment]
thanks for looking at it!
oh btw: is there any option to have the effect of additionalProperties true at the beginning of the generated code, i.e. get this instead:
```typescript
export interface ExampleSchema {
  [k: string]: unknown;
  firstName: string;
  lastName: string;
  /**
   * Age in years
   */
  age?: number;
  height?: number;
  favoriteFoods?: unknown[];
  likesDogs?: boolean;
}
```

**Comment by bcherny** [comment]
No, but feel free to open an issue if you’d like to gauge interest.

**Comment by john30** [comment]
note: also addressed by #383
